# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/cocoa/nsTouchBar.mm
# Commit: 64d7f0db63a4
# Full Hash: 64d7f0db63a45ec431847e77c6220f5d2ef9c4a9
# Author: harry <htwyford@mozilla.com>
# Date: 2019-10-11 09:30:10
# Regressor Bug: 1563350
# File Overlap Count: 2
# Description:
#   Bug 1563350 - Add popover to the Touch Bar that displays when the Urlbar has focus. r=mikedeboer,spohl,fluent-reviewers,Pike,flod
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D38563
# ==============================================================================

diff -r 879c857b5b65 -r 64d7f0db63a4 widget/cocoa/nsTouchBar.mm
--- a/widget/cocoa/nsTouchBar.mm	Wed Oct 09 10:46:05 2019 +0000
+++ b/widget/cocoa/nsTouchBar.mm	Thu Oct 10 19:57:32 2019 +0000
@@ -19,6 +19,12 @@
 static NSTouchBarItemIdentifier ShareScrubberIdentifier =
     [TouchBarInput nativeIdentifierWithType:@"scrubber" withKey:@"share"];
 
+// The search popover needs to show/hide depending on if the Urlbar is focused
+// when it is created. We keep track of its identifier to accomodate this
+// special handling.
+static NSTouchBarItemIdentifier SearchPopoverIdentifier =
+    [TouchBarInput nativeIdentifierWithType:@"popover" withKey:@"search-popover"];
+
 // Used to tie action strings to buttons.
 static char sIdentifierAssociationKey;
 
@@ -97,7 +103,7 @@
         [TouchBarInput nativeIdentifierWithType:@"button" withKey:@"reload"],
         [TouchBarInput nativeIdentifierWithType:@"mainButton" withKey:@"open-location"],
         [TouchBarInput nativeIdentifierWithType:@"button" withKey:@"new-tab"],
-        ShareScrubberIdentifier
+        ShareScrubberIdentifier, SearchPopoverIdentifier
       ];
       self.defaultItemIdentifiers = [defaultItemIdentifiers copy];
     } else {
@@ -353,6 +359,15 @@
   } else {
     aPopoverItem.collapsedRepresentation = nil;
   }
+
+  // Special handling to show/hide the search popover if the Urlbar is focused.
+  if ([[aInput nativeIdentifier] isEqualToString:SearchPopoverIdentifier]) {
+    bool urlbarIsFocused = false;
+    mTouchBarHelper->GetIsUrlbarFocused(&urlbarIsFocused);
+    if (urlbarIsFocused) {
+      [aPopoverItem showPopover:self];
+    }
+  }
 }
 
 - (void)updateScrollView:(NSCustomTouchBarItem*)aScrollViewItem input:(TouchBarInput*)aInput {
@@ -483,7 +498,10 @@
       continue;
     }
 
-    if ([[input type] hasSuffix:@"popover"]) {
+    // Childless popovers contain the default Touch Bar as its popoverTouchBar.
+    // We check for [input children] since the default Touch Bar contains a
+    // popover (search-popover), so this would infinitely loop if there was no check.
+    if ([[input type] hasSuffix:@"popover"] && [input children]) {
       NSTouchBarItem* item = [self itemForIdentifier:identifier];
       [(nsTouchBar*)[(NSPopoverTouchBarItem*)item popoverTouchBar] releaseJSObjects];
     }